### PR TITLE
(3.0.6 backport) CBG-2842 Always return unauthorized for non authorized public API (#6183)

### PR DESCRIPTION
**Pull request by torcolvin**
CBG-2850

Backport of https://github.com/couchbase/sync_gateway/commit/e0520bcbe86c5d12a35dba8a77bc73f15e09fa6b

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration-go1.16.15/42/

### CODE REVIEW

File: rest/handler.go [hunk 1]
```diff
@@ -209,17 +213,24 @@ func (h *handler) invoke(method handlerMethod, accessPermissions []Permission, r
...
+					if !authorized || h.providedAuthCredentials() {
+						return ErrInvalidLogin
+					}
+				} else {
+					if h.privs == regularPrivs || h.privs == publicPrivs {
+						if !h.providedAuthCredentials() {
+
+							return ErrLoginRequired
+						}
+						return ErrInvalidLogin
```
Comment (bbrks) [inline rest/handler.go]: I'm not sure why this change is different than the one that went into `master` ?

https://github.com/couchbase/sync_gateway/pull/6183/files#diff-9899f6c02129d809f5816c71f72902fd77314b35b000f7484d4ba6ea2e5ca1ef

File: rest/session_api.go
Comment (bbrks) [inline rest/session_api.go:1]: No `session_test.go` change for `TestSessionExtension` error message ?

File: rest/admin_api_auth_test.go [hunk 2]
```diff
@@ -508,6 +508,7 @@ func TestAdminAuthWithX509(t *testing.T) {
 	_, statusCode, err = checkAdminAuth("", "invalidUser", "invalidPassword", "", httpClient, managementEndpoints, true, []Permission{{"!admin", false}}, nil)
 	assert.NoError(t, err)
 	assert.Equal(t, http.StatusUnauthorized, statusCode)
+	require.Contains(t, err.Error(), ErrInvalidLogin.Message)
 }
 
 func TestAdminAPIAuth(t *testing.T) {
```
Comment (bbrks) [inline rest/admin_api_auth_test.go:514]: There a new assertion for error message in this test from the original PR in this test

File: rest/admin_api_test.go [hunk 3]
```diff
@@ -3635,11 +3635,13 @@ func TestNotExistentDBRequest(t *testing.T) {
 
 	// Request to non-existent db with valid credentials
 	resp := rt.SendAdminRequestWithAuth("PUT", "/dbx/_config", "", "random", "password")
-	assertStatus(t, resp, http.StatusForbidden)
+	assertStatus(t, resp, http.StatusUnauthorized)
+	require.Contains(t, resp.Body.String(), ErrInvalidLogin.Message)
```
Comment (bbrks) [inline rest/admin_api_test.go]: This is different than the status code assertion on `master`, which is still `Forbidden`

File: rest/api_test.go [hunk 4]
```diff
@@ -118,10 +124,15 @@ func TestDisablePublicBasicAuth(t *testing.T) {
...
-
+	require.Contains(t, response.Body.String(), ErrInvalidLogin.Message)
 	// Also check that we can't create a session through POST /db/_session
 	response = rt.SendRequest(http.MethodPost, "/db/_session", `{"name":"user1","password":"letmein"}`)
 	assertStatus(t, response, http.StatusUnauthorized)
+	require.Contains(t, response.Body.String(), ErrLoginRequired.Message)
+
+	response = rt.SendRequest(http.MethodPost, "/notadb/_session", `{"name":"user1","password":"letmein"}`)
+	assertStatus(t, response, http.StatusUnauthorized)
+	require.Contains(t, response.Body.String(), ErrLoginRequired.Message)
```
Comment (bbrks) [inline rest/api_test.go:141]: Nit: This is above the assertions for`/_session` in `master`. It would probably be best to keep them consistent.

**Review by bbrks — CHANGES_REQUESTED** [review]
This looks slightly different than the code that went into `master` - is that intentional?

File: rest/admin_api_test.go (hunk 3 above)
Comment (torcolvin) [reply rest/admin_api_test.go:3639]: > This looks slightly different than the code that went into `master` - is that intentional?

I wish it weren't different, but the code that handles inactiveDatabases makes the error codes different. As long as the error codes match for existing database but unable to access and non existent database, I think this is OK.

The starting state of 3.0 and 3.1 was always different https://docs.google.com/spreadsheets/d/1aXL9D7R54R0prYr-6ini-Qrk3Pj42k3gE4VQbP-lCaI/edit#gid=0 and I've preserved the error codes.

**Comment by torcolvin** [comment]
> This looks slightly different than the code that went into `master` - is that intentional?

I wish it weren't different, but the code that handles inactiveDatabases makes the error codes different. As long as the error codes match for existing database but unable to access and non existent database, I think this is OK.

The starting state of 3.0 and 3.1 was always different https://docs.google.com/spreadsheets/d/1aXL9D7R54R0prYr-6ini-Qrk3Pj42k3gE4VQbP-lCaI/edit#gid=0 and I've preserved the error codes.

File: rest/handler.go [hunk 5]
```diff
@@ -209,17 +213,24 @@ func (h *handler) invoke(method handlerMethod, accessPermissions []Permission, r
...
+			if httpError, ok := err.(*base.HTTPError); ok && httpError.Status == http.StatusNotFound {
+				if shouldCheckAdminAuth {
 					authorized, err := h.checkAdminAuthenticationOnly()
 					if err != nil {
 						return err
 					}
 
-					if authorized {
-						return base.HTTPErrorf(http.StatusForbidden, "")
+					if !authorized || h.providedAuthCredentials() {
```
Comment (adamcfraser) [inline rest/handler.go]: On master we only use providedAuthCredentials on the public port case - it looks like we're using it in conjunction with admin auth here (inside shouldCheckAdminAuth), which seems incorrect?  Which combination is this handling?

File: rest/handler.go [hunk 6]
```diff
@@ -209,17 +213,24 @@ func (h *handler) invoke(method handlerMethod, accessPermissions []Permission, r
...
+						return ErrInvalidLogin
+					}
```
Comment (adamcfraser) [inline rest/handler.go:225]: Line 225 looks like it should have `else { return base.HTTPErrorf(http.StatusForbidden, "")}` to handle the case of an admin user that successfully authenticated requesting a db that doesn't exist.

File: rest/handler.go [hunk 7]
```diff
@@ -209,17 +213,24 @@ func (h *handler) invoke(method handlerMethod, accessPermissions []Permission, r
...
+				} else {
+					if h.privs == regularPrivs || h.privs == publicPrivs {
```
Comment (adamcfraser) [inline rest/handler.go:228]: Is this privs check required, if we're in the else for "shouldCheckAdminAuth"?

**Review by adamcfraser — DISMISSED** [review]

File: rest/handler.go (hunk 7 above)
Comment (torcolvin) [reply rest/handler.go:228]: This covers the case (which is maybe test-only) where h.server.config.API.AdminInterfaceAuthentication is nil

Comment (adamcfraser) [reply rest/handler.go:228]: I see.  These conditions are organized in a way that's a bit hard to follow, but I guess the `else` on line 226 covers both "public port" and "admin port but admin auth is disabled", and the latter falls through to returning 'not found' on line 237.

File: rest/handler.go (hunk 5 above)
Comment (torcolvin) [reply rest/handler.go]: Seems to be unused.

File: rest/handler.go (hunk 6 above)
Comment (torcolvin) [reply rest/handler.go:225]: Fixed.

**Comment by adamcfraser** [comment]
This looks good to me, and integration test is clean.  @bbrks can you sanity check that your review comments are addressed before we merge this?

**Review by bbrks — APPROVED** [review]
